### PR TITLE
Define the onErrorResume stage.

### PR DESCRIPTION
**Pull request by cescoffier**
This stage is the first error recovery stage. It allows supplying a value when a stream encounters an error.

### CODE REVIEW

File: streams/tck/src/main/java/org/eclipse/microprofile/reactive/streams/tck/OnErrorResumeStageVerification.java [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+          throw new RuntimeException("failed");
+        })
+        .toList()
+        .run(getEngine()));
+  }
+
+
+  @Override
+  List<Object> reactiveStreamsTckVerifiers() {
+    return Collections.emptyList();
```
Comment (cescoffier) [inline streams/tck/src/main/java/org/eclipse/microprofile/reactive/streams/tck/OnErrorResumeStageVerification.java]: I would be great to have a way to test the compliance of the stage against the reactive stream TCK. However, I didn't find a way. @jroper do you have an idea?

File: streams/tck/src/main/java/org/eclipse/microprofile/reactive/streams/tck/OnErrorResumeStageVerification.java [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+        return "foo";
+      })
+      .toList()
+      .run(getEngine())), Arrays.asList("A", "foo"));
+    assertEquals(exception.get().getMessage(), "failed");
+  }
+
+
+  @Test(expectedExceptions = RuntimeException.class)
+  public void onErrorResumeStageShouldPropagateRuntimeExceptions() {
```
Comment (cescoffier) [inline streams/tck/src/main/java/org/eclipse/microprofile/reactive/streams/tck/OnErrorResumeStageVerification.java]: RX and Reactor throw a composite exception in this case. This composite exception encapsulates the source exception and the "failed" exception. It means we can't test for the message of the thrown exception as it's something like: Composite Exception encountered....

Comment (jroper) [reply streams/tck/src/main/java/org/eclipse/microprofile/reactive/streams/tck/OnErrorResumeStageVerification.java]: Fair enough, I don't think we have to be too strict about specifying exactly what error is thrown in this circumstance.

File: streams/tck/src/main/java/org/eclipse/microprofile/reactive/streams/tck/OnErrorResumeStageVerification.java [hunk 3]
```diff
@@ -0,0 +1,192 @@
...
+  public void onErrorResumeWithPublisherShouldBeAbleToInjectAFailure() {
+    await(ReactiveStreams.failed(new Exception("failed"))
+      .onErrorResumeWithPublisher(err -> ReactiveStreams.failed(new Exception("boom")).buildRs())
+      .toList()
+      .run(getEngine()));
+  }
+
+  @Override
+  List<Object> reactiveStreamsTckVerifiers() {
+    return Collections.emptyList();
```
Comment (jroper) [inline streams/tck/src/main/java/org/eclipse/microprofile/reactive/streams/tck/OnErrorResumeStageVerification.java]: Could we return a processor verifier for both resume and resume with that simply rethrew the exception? Also, a resume with publisher verifier that started with a failed publisher, and then resumed with the actual publisher would be good. These will cover all the cases like ensuring that if no exception is thrown, it completes normally, etc.

File: streams/tck/src/main/java/org/eclipse/microprofile/reactive/streams/tck/OnErrorResumeStageVerification.java (hunk 2 above)
Comment (cescoffier) [reply streams/tck/src/main/java/org/eclipse/microprofile/reactive/streams/tck/OnErrorResumeStageVerification.java]: Well, I have been thinking about it, and I don't really like this `CompositeException`, the error passed to the `onError` should be the error thrown by the function. I've updated the code to check it.

File: streams/tck/src/main/java/org/eclipse/microprofile/reactive/streams/tck/OnErrorResumeStageVerification.java [hunk 4]
```diff
@@ -183,10 +187,95 @@ public void onErrorResumeWithPublisherShouldBeAbleToInjectAFailure() {
...
+        }
+
+        @Override
+        public Integer createElement(int element) {
+            return element;
+        }
+
+        @Override
+        public void required_spec203_mustNotCallMethodsOnSubscriptionOrPublisherInOnError() {
+            // Ignore this test, as the subscription happen in the onError method to switch the control to the new stream.
```
Comment (cescoffier) [inline streams/tck/src/main/java/org/eclipse/microprofile/reactive/streams/tck/OnErrorResumeStageVerification.java]: @jroper this is a bit annoying. What do you think?

Comment (jroper) [reply streams/tck/src/main/java/org/eclipse/microprofile/reactive/streams/tck/OnErrorResumeStageVerification.java]: Looks sus - sounds like either rxjava or the `ConnectableProcessor` is doing something wrong here if this is failing. Let me know if you'd like me to take a look.

**Comment by jroper** [comment]
One question, in `onErrorResumeWith`, if the returned publisher emits an error, should `onErrorResumeWith` be invoked again to handle it? Akka streams, by default, does this indefinitely, it also has a `recoverWithRetries` variant to control how many times it will retry (passing 1 to that implements the behaviour as currently specified).

Which one makes sense really depends on the use case. If I'm using this to restart a stream in the event of failure, then I want the retries. If however I'm using this to switch to an alternate stream in the event of failure, then I don't want retries.

It's possible to implement both with either behaviour, if I want to keep retrying and the API offers the non retry version, then I just need to return a publisher that has the retry it (using recursion to create it). If I want to retry a limited number of times, I simply take in a retries parameter in my recursive builder and decrement it. If however retries are automatic, then to not retry I need to maintain some state, eg a count down or a boolean, and when I don't want to retry anymore, I can either return a completed publisher, or throw the exception if I wanted it propagated. I think the former is easier/cleaner, so potentially the best option is to offer the not retrying option.

The docs should make it clear which behaviour is done though.

**Comment by cescoffier** [comment]
The documentation of the stage right now is around these lines:

```
 Returns a stream containing all the elements from this stream. Additionally, in the case of failure, rather than invoking `#onError(Consumer)`, it invokes the given method and emits the returned `PublisherBuilder` instead.
     
By default, when a stream encounters an error that prevents it from emitting the expected item to its subscriber, the stream (publisher) invokes its subscriber's `onError` method, and then terminate without invoking any more of its subscriber's methods. This operator changes this behavior. If the current stream encounters an error, instead of invoking its subscriber's `onError` method, it will instead relinquish control to the `PublisherBuilder` returned from given function, which invokes the subscriber's `onNext` method if it is able to do so. In such a case, because no publisher necessarily invokes `onError`, the subscriber may never know that an error happened.
```

So, if the returned stream emits a failure the `onErrorResumeWith` won't be called again as the returned stream is used as _source_ for the rest of the pipeline starting just after the `onResumeWith` stage. That's, obviously, debatable. I actually think that retries should be covered in another issue as they are pretty important but can be hard to handle especially in a streaming application using "windows of data".

**Pull request merged**